examples/step-55/step-55.cc

```diff
@@ -0,0 +1,1181 @@
+  template <int dim>
+  class RightHandSide : public Function<dim>
+  {
+  public:
+    RightHandSide () : Function<dim>(dim+1) {}
+
+    virtual double value (const Point<dim> &p,
+                          const unsigned int component = 0) const;
+  };
+
+  template <>
+  double
+  RightHandSide<2>::value (const Point<2> &p,
+                           const unsigned int component) const
+  {
+    using numbers::PI;
+    double x = p(0);
+    double y = p(1);
+    if (component == 0)
+      return PI * PI * sin(PI * x) + PI * cos(PI * x) * cos(PI * y);
+    if (component == 1)
+      return - PI * PI * PI * y * cos(PI * x) - PI * sin(PI * y) * sin(PI * x);
+    if (component == 2)
+      return 0;
+
+    Assert (false, ExcMessage ("Component out of range in RightHandSide"));
+    return 0;
+
+  }
+
+  template <>
+  double
+  RightHandSide<3>::value (const Point<3>   &p,
+                           const unsigned int component) const
+  {
+    using numbers::PI;
+    double x = p(0);
+    double y = p(1);
+    double z = p(2);
+    if (component == 0)
+      return 2 * PI * PI * sin(PI * x) + PI * cos(PI * x) * cos(PI * y) * sin(PI * z);
+    if (component == 1)
+      return  - PI * PI * PI * y * cos (PI * x) + PI * (-1) * sin(PI * y)*sin(PI * x)*sin(PI * z);
+    if (component == 2)
+      return - PI * PI * PI * z * cos (PI * x) + PI * cos(PI * z)*sin(PI * x)*cos(PI * y);
+    if (component == 3)
+      return 0;
+
+    Assert (false, ExcMessage ("Component out of range in RightHandSide"));
+    return 0;
+  }
+
+  // Sadly, we need a separate function for the boundary conditions
+  // to be used in the geometric multigrid. This is because it needs
+  // to be a function with $dim$ components, whereas Solution has
+  // $dim+1$ components. Rather than copying the implementation of
+  // Solution, we forward the calls to Solution::value. For that we need
+  // an instance of the class Solution, which you can find as a private
+  // member.
+  template <int dim>
+  class BoundaryValuesForVelocity : public Function<dim>
+  {
+  public:
+    BoundaryValuesForVelocity () : Function<dim>(dim) {}
+
+    virtual double value (const Point<dim>   &p,
+                          const unsigned int  component = 0) const;
+
+  private:
+    Solution<dim> solution;
+  };
+
+
+  template <int dim>
+  double
+  BoundaryValuesForVelocity<dim>::value (const Point<dim>  &p,
+                                         const unsigned int component) const
+  {
+    Assert (component < this->n_components,
+            ExcIndexRange (component, 0, this->n_components));
+
+    return solution.value(p, component);
+  }
+
+
+
+  // @sect3{ASPECT BlockSchurPreconditioner}
+
+  // This class, which is taken from ASPECT and then slightly modified,
+  // implements the block Schur preconditioner for the Stokes system discussed
+  // above. It is templated on the types
+  // for the preconditioner blocks for velocity and schur complement.
```

Comment:
I would say make it similar to the step-22 code. We're in the *tutorial* here, and it makes more sense to have backward references to simpler tutorial programs, than to reference something they could look up in a code that is probably 2 orders of magnitude larger and more complex than the current code.

It *is* ok to say something like this maybe: "In the following, we will implement a preconditioner that expands a little bit on the ideas discussed in the Results section of step-22. Specifically, we will do X, Y, and Z. A further iteration of this approach is also used in the ASPECT code (see http://aspect.dealii.org) that solves the Stokes equations in the context of simulating convection in the earth mantle, and which has been used to solve problems on many thousands of processors."

This provides more of a "if you're curious about more, go there", rather than "if you want to understand what we're doing here, you need to read what's done there".